From 75eb0a3f6f7a1cc89aad5ee0a6616cbb912b770c Mon Sep 17 00:00:00 2001
From: SamLukeYes <samlukeyes123@gmail.com>
Date: Thu, 20 Apr 2023 18:27:52 +0800
Subject: [PATCH] qq: add missing dependencies

---
 .../instant-messengers/qq/default.nix           | 17 ++++++++++++++---
 1 file changed, 14 insertions(+), 3 deletions(-)

diff --git a/pkgs/applications/networking/instant-messengers/qq/default.nix b/pkgs/applications/networking/instant-messengers/qq/default.nix
index 32d34787d6579..9f037845def66 100644
--- a/pkgs/applications/networking/instant-messengers/qq/default.nix
+++ b/pkgs/applications/networking/instant-messengers/qq/default.nix
@@ -6,6 +6,7 @@
 , glib
 , gtk3
 , lib
+, libappindicator
 , libdrm
 , libgcrypt
 , libkrb5
@@ -14,6 +15,7 @@
 , xorg
 , systemd
 , stdenv
+, vips
 , at-spi2-core
 , autoPatchelfHook
 , wrapGAppsHook
@@ -49,18 +51,20 @@ stdenv.mkDerivation {
     alsa-lib
     at-spi2-core
     cups
-    gtk3
     glib
+    gtk3
     libdrm
     libgcrypt
     libkrb5
     mesa
     nss
+    vips
     xorg.libXdamage
   ];
 
-  runtimeDependencies = [
-    (lib.getLib systemd)
+  runtimeDependencies = map lib.getLib [
+    libappindicator
+    systemd
   ];
 
   installPhase = ''
@@ -74,9 +78,16 @@ stdenv.mkDerivation {
       --replace "/usr/share" "$out/share"
     ln -s $out/opt/QQ/qq $out/bin/qq
 
+    # Remove bundled libraries
+    rm -r $out/opt/QQ/resources/app/sharp-lib
+
     runHook postInstall
   '';
 
+  preFixup = ''
+    gappsWrapperArgs+=(--prefix PATH : "${lib.makeBinPath [ gjs ]}")
+  '';
+
   meta = with lib; {
     homepage = "https://im.qq.com/linuxqq/";
     description = "Messaging app";
