diff --git a/pkgs/by-name/wa/waydroid-helper/package.nix b/pkgs/by-name/wa/waydroid-helper/package.nix
index 42dd8bda565a..6df9a186df4a 100644
--- a/pkgs/by-name/wa/waydroid-helper/package.nix
+++ b/pkgs/by-name/wa/waydroid-helper/package.nix
@@ -16,8 +16,10 @@
   dbus,
   systemd,
   bash,
+  e2fsprogs,
   fakeroot,
   gobject-introspection,
+  unzip,
 }:
 
 python3Packages.buildPythonApplication rec {
@@ -38,7 +40,9 @@ python3Packages.buildPythonApplication rec {
       --replace-fail "dbus_service_dir," "'$out/share/dbus-1/system-services',"
     substituteInPlace systemd/meson.build \
       --replace-fail ": systemd_system_unit_dir" ": '$out/lib/systemd/system'" \
-      --replace-fail ": systemd_user_unit_dir" ": '$out/lib/sysusers.d'"
+      --replace-fail ": systemd_user_unit_dir" ": '$out/lib/systemd/user'"
+    substituteInPlace systemd/{system/waydroid-mount,user/waydroid-monitor}.service \
+      --replace-fail "/usr/bin/waydroid-helper" "$out/bin/waydroid-helper"
     # com.jaoushingan.WaydroidHelper.desktop: component-name-missing, description-first-para-too-short
     # url-homepage-missing, desktop-app-launchable-omitted, content-rating-missing, developer-info-missing
     sed -i '/test(/{N;/Validate appstream file/!b;:a;N;/)/!ba;d}' data/meson.build
@@ -80,7 +84,11 @@ python3Packages.buildPythonApplication rec {
 
   makeWrapperArgs = [
     "\${gappsWrapperArgs[@]}"
-    "--prefix PATH : ${lib.makeBinPath [ fakeroot ]}"
+    "--prefix PATH : ${lib.makeBinPath [
+      e2fsprogs
+      fakeroot
+      unzip
+    ]}"
   ];
 
   postInstallCheck = ''
